src/Storage/StreamWrapper.php

```diff
@@ -0,0 +1,167 @@
+<?php
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage;
+
+use Google\Cloud\Exception\GoogleException;
+
+/**
+ * A streamWrapper implementation for handling `gs://bucket/path/to/file.jpg`
+ *
+ * See: http://php.net/manual/en/class.streamwrapper.php
+ */
+class StreamWrapper
+{
+    // Must be public according to the PHP documentation
+    public $context;
+
+    // a GuzzleHttp\Psr7\StreamInterface instance
+    private $stream;
+
+    private $protocol;
+    private $bucket;
+    private $file;
+    private $mode;
+    private $options;
+
+    public function __destruct()
+    {
+        $this->stream_close();
+    }
+
+    // @codingStandardsIgnoreStart
+    public function stream_open($path, $mode, $options, &$opened_path)
+    {
+        // @codingStandardsIgnoreEnd
+        $url = parse_url($path);
+        $this->protocol = $url['scheme'];
+        $this->file = substr($url['path'], 1);
+        $this->mode = $mode;
+
+        $client = $this->getOption('client') ?: new StorageClient();
```

Comment:
The `getOption('client')` is retrieving an optional client from the stream context and falling back to the default StorageClient you get from ADC. I could use a static method to set the client to use as well.